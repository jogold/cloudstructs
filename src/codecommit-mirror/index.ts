import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import * as codecommit from 'aws-cdk-lib/aws-codecommit';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

/**
 * Properties for a CodeCommitMirror
 */
export interface CodeCommitMirrorProps {
  /**
   * The source repository
   */
  readonly repository: CodeCommitMirrorSourceRepository;

  /**
   * The ECS cluster where to run the mirroring operation
   */
  readonly cluster: ecs.ICluster;

  /**
   * The schedule for the mirroring operation
   *
   * @default - everyday at midnight
   */
  readonly schedule?: events.Schedule;

  /**
   * Where to run the mirroring Fargate tasks
   *
   * @default - public subnets
   */
  readonly subnetSelection?: ec2.SubnetSelection;
}

/**
 * A source repository for AWS CodeCommit mirroring
 */
export abstract class CodeCommitMirrorSourceRepository {
  /**
   * Public GitHub repository
   */
  public static gitHub(owner: string, name: string): CodeCommitMirrorSourceRepository {
    return {
      name,
      plainTextUrl: `https://github.com/${owner}/${name}`,
    };
  }

  /**
   * Private repository with HTTPS clone URL stored in a AWS Secrets Manager secret or
   * a AWS Systems Manager secure string parameter.
   *
   * @param name the repository name
   * @param url the secret containing the HTTPS clone URL
   */
  public static private(name: string, url: ecs.Secret): CodeCommitMirrorSourceRepository {
    return {
      name,
      secretUrl: url,
    };
  }

  /**
   * The name of the repository
   */
  public abstract readonly name: string;

  /** The HTTPS clone URL in plain text, used for a public repository */
  public abstract readonly plainTextUrl?: string;

  /**
   * The HTTPS clone URL if the repository is private.
   *
   * The secret should contain the username and/or token.
   *
   * @example
   * `https://TOKEN@github.com/owner/name`
   * `https://USERNAME:TOKEN@bitbucket.org/owner/name.git`
   */
  public abstract readonly secretUrl?: ecs.Secret;
}

/**
 * Mirror a repository to AWS CodeCommit on schedule
 */
export class CodeCommitMirror extends Construct {
  constructor(scope: Construct, id: string, props: CodeCommitMirrorProps) {
    super(scope, id);

    const destination = new codecommit.Repository(this, 'Repository', {
      repositoryName: props.repository.name,
      description: `Mirror of ${props.repository.name}`,
    });

    const taskDefinition = new ecs.FargateTaskDefinition(this, 'TaskDefinition');

    taskDefinition.addContainer('Container', {
      image: ecs.ContainerImage.fromAsset(path.join(__dirname, '..', '..', 'assets', 'codecommit-mirror', 'docker')),
      logging: new ecs.AwsLogDriver({
        streamPrefix: props.repository.name,
        logRetention: logs.RetentionDays.TWO_MONTHS,
      }),
      environment: {
        NAME: props.repository.name,
        DESTINATION: `codecommit::${cdk.Stack.of(destination).region}://${destination.repositoryName}`,
        ...props.repository.plainTextUrl
          ? { SOURCE: props.repository.plainTextUrl }
          : {},
      },
      secrets: props.repository.secretUrl
        ? { SOURCE: props.repository.secretUrl }
        : undefined,
    });

    taskDefinition.addToTaskRolePolicy(new iam.PolicyStatement({
      actions: ['codecommit:GitPush'],
      resources: [destination.repositoryArn],
    }));

    const rule = new events.Rule(this, 'Rule', {
      schedule: props.schedule ?? events.Schedule.cron({
        minute: '0',
        hour: '0',
      }),
    });

    rule.addTarget(new targets.EcsTask({
      cluster: props.cluster,
      taskDefinition,
      subnetSelection: props.subnetSelection ?? { subnetType: ec2.SubnetType.PUBLIC },
    }));
  }
}
