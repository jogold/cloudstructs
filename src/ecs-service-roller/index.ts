import type { UpdateServiceCommandInput } from '@aws-sdk/client-ecs';
import { Stack } from 'aws-cdk-lib';
import * as ecs from 'aws-cdk-lib/aws-ecs';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * Properties for a EcsServiceRoller
 */
export interface EcsServiceRollerProps {
  /**
   * The ECS cluster where the services run
   */
  readonly cluster: ecs.ICluster;

  /**
   * The ECS service for which tasks should be rolled
   */
  readonly service: ecs.IService;

  /**
   * The rule or schedule that should trigger a roll
   *
   * @default - roll everyday at midnight
   */
  readonly trigger?: RollTrigger;
}

/**
 * The rule or schedule that should trigger a roll
 */
export abstract class RollTrigger {
  /**
   * Schedule that should trigger a roll
   */
  public static fromSchedule(schedule: events.Schedule): RollTrigger {
    return { schedule };
  }

  /**
   * Rule that should trigger a roll
   */
  public static fromRule(rule: events.Rule): RollTrigger {
    return { rule };
  }

  /**
   * Roll schedule
   *
   * @default - roll everyday at midnight
   */
  public abstract readonly schedule?: events.Schedule;

  /**
   * Roll rule
   *
   * @default - roll everyday at midnight
   */
  public abstract readonly rule?: events.Rule;
}

/**
 * Roll your ECS service tasks on schedule or with a rule
 */
export class EcsServiceRoller extends Construct {
  constructor(scope: Construct, id: string, props: EcsServiceRollerProps) {
    super(scope, id);

    const rule = props.trigger?.rule ?? new events.Rule(this, 'Rule', {
      schedule: props.trigger?.schedule ?? events.Schedule.cron({
        minute: '0',
        hour: '0',
      }),
    });


    const parameters: UpdateServiceCommandInput = {
      service: props.service.serviceName,
      cluster: props.cluster.clusterName,
      forceNewDeployment: true,
    };
    rule.addTarget(new targets.AwsApi({
      service: 'ECS',
      action: 'updateService',
      parameters,
      // https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-supported-iam-actions-resources.html
      // arn:aws:ecs:<region>:<account>:service/<cluster-name>/<service-name>
      policyStatement: new iam.PolicyStatement({
        actions: ['ecs:UpdateService'],
        resources: [Stack.of(this).formatArn({
          service: 'ecs',
          resource: 'service',
          resourceName: `${props.cluster.clusterName}/${props.service.serviceName}`,
        })],
      }),
    }));
  }
}
