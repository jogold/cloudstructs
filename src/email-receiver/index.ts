export * from './receiver';
export * from './types';
