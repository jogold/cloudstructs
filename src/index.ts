export * from './slack-events';
export * from './slack-textract';
export * from './static-website';
export * from './state-machine-cr-provider';
export * from './email-receiver';
export * from './ecs-service-roller';
export * from './url-shortener';
export * from './saml-identity-provider';
export * from './codecommit-mirror';
export * from './slack-app';
export * from './toolkit-cleaner';
export * from './ssl-server-test';
