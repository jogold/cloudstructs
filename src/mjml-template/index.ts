import { CfnTemplate } from 'aws-cdk-lib/aws-ses';
import { Construct } from 'constructs';
import mjml2html = require('mjml'); // eslint-disable-line @typescript-eslint/no-require-imports

/**
 * Properties for a MjmlTempalte
 */
export interface MjmlTemplateProps {
  /**
   * The name of the template
   *
   * @default - a CloudFormation generated name
   */
  readonly templateName?: string;

  /**
   * The subject line of the email
   */
  readonly subject: string;

  /**
   * The MJML of the email
   */
  readonly mjml: string;
}

/**
 * SES email template from [MJML](https://mjml.io/)
 */
export class MjmlTemplate extends Construct {
  /**
   * The name of the template
   */
  public readonly templateName: string;

  constructor(scope: Construct, id: string, props: MjmlTemplateProps) {
    super(scope, id);

    const template = new CfnTemplate(this, 'Resource', {
      template: {
        templateName: props.templateName,
        subjectPart: props.subject,
        htmlPart: mjml2html(props.mjml).html,
      },
    });

    this.templateName = template.attrId;
  }
}