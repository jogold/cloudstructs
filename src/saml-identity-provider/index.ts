import type { CreateSAMLProviderCommandInput, UpdateSAMLProviderCommandInput, DeleteSAMLProviderCommandInput } from '@aws-sdk/client-iam';
import { Names, Stack } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as cr from 'aws-cdk-lib/custom-resources';
import { Construct } from 'constructs';

/**
 * Properties for a SamlProvider
 *
 * @deprecated use `SamlProviderProps` from `aws-cdk-lib/aws-iam`
 */
export interface SamlIdentityProviderProps {
  /**
   * A name for the SAML identity provider
   *
   * @default - derived for the node's unique id
   */
  readonly name?: string;

  /**
   * An XML document generated by an identity provider (IdP) that supports SAML 2.0.
   *
   * The document includes the issuer's name, expiration information, and keys that
   * can be used to validate the SAML authentication response (assertions) that are
   * received from the IdP. You must generate the metadata document using the identity
   * management software that is used as your organization's IdP.
   */
  readonly metadataDocument: string;
}

/**
 * Create a SAML identity provider
 *
 * @deprecated use `SamlProvider` from `aws-cdk-lib/aws-iam`
 */
export class SamlIdentityProvider extends Construct {
  /**
   * The ARN of the SAML identity provider
   */
  public readonly samlIdentityProviderArn: string;

  constructor(scope: Construct, id: string, props: SamlIdentityProviderProps) {
    super(scope, id);

    const name = props.name ?? `${Names.uniqueId(this)}IdentityProvider`;

    const arn = Stack.of(this).formatArn({
      service: 'iam',
      region: '',
      resource: 'saml-provider',
      resourceName: name,
    });

    const idp = new cr.AwsCustomResource(this, 'Resource', {
      resourceType: 'Custom::SamlIdentityProvider',
      onCreate: {
        service: 'IAM',
        action: 'createSAMLProvider',
        parameters: {
          Name: name,
          SAMLMetadataDocument: props.metadataDocument,
        } as CreateSAMLProviderCommandInput,
        physicalResourceId: cr.PhysicalResourceId.fromResponse('SAMLProviderArn'),
      },
      onUpdate: {
        service: 'IAM',
        action: 'updateSAMLProvider',
        parameters: {
          SAMLProviderArn: new cr.PhysicalResourceIdReference().toJSON(),
          SAMLMetadataDocument: props.metadataDocument,
        } as UpdateSAMLProviderCommandInput,
        physicalResourceId: cr.PhysicalResourceId.fromResponse('SAMLProviderArn'),
      },
      onDelete: {
        service: 'IAM',
        action: 'deleteSAMLProvider',
        parameters: {
          SAMLProviderArn: new cr.PhysicalResourceIdReference().toJSON(),
        } as DeleteSAMLProviderCommandInput,
      },
      policy: cr.AwsCustomResourcePolicy.fromStatements([
        new iam.PolicyStatement({
          actions: [
            'iam:createSAMLProvider',
            'iam:updateSAMLProvider',
            'iam:deleteSAMLProvider',
          ],
          resources: [arn],
        }),
      ]),
    });

    this.samlIdentityProviderArn = idp.getResponseField('SAMLProviderArn');
  }
}

/**
 * Principal entity that represents a SAML federated identity provider.
 *
 * @deprecated use `SamlPrincipal` from `aws-cdk-lib/aws-iam`
 */
export class SamlFederatedPrincipal extends iam.FederatedPrincipal {
  constructor(identityProvider: SamlIdentityProvider) {
    super(
      identityProvider.samlIdentityProviderArn,
      {
        StringEquals: {
          'SAML:aud': 'https://signin.aws.amazon.com/saml',
        },
      },
      'sts:AssumeRoleWithSAML',
    );
  }
}
