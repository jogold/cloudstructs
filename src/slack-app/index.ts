export * from './slack-app';
export * from './manifest';
