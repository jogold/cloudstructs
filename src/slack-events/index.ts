import * as cdk from 'aws-cdk-lib';
import * as apigatewayv2 from 'aws-cdk-lib/aws-apigatewayv2';
import * as integrations from 'aws-cdk-lib/aws-apigatewayv2-integrations';
import * as events from 'aws-cdk-lib/aws-events';
import * as logs from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';
import { EventsFunction } from './events-function';

/**
 * Properties for a SlackEvents
 */
export interface SlackEventsProps {
  /**
   * The signing secret of the Slack app
   */
  readonly signingSecret: cdk.SecretValue;

  /**
   * A name for the API Gateway resource
   *
   * @default SlackEventsApi
   */
  readonly apiName?: string;

  /**
   * Whether to use a custom event bus
   *
   * @default false
   */
  readonly customEventBus?: boolean;
}

/**
 * Send Slack events to Amazon EventBridge
 */
export class SlackEvents extends Construct {
  /**
   * The custom event bus where Slack events are sent
   */
  public readonly eventBus?: events.EventBus;

  constructor(scope: Construct, id: string, props: SlackEventsProps) {
    super(scope, id);

    if (props.customEventBus) {
      this.eventBus = new events.EventBus(this, 'EventBus');
    }

    // Send event to the event bus
    const handler = new EventsFunction(this, 'handler', {
      logRetention: logs.RetentionDays.ONE_MONTH,
      environment: {
        SLACK_SIGNING_SECRET: props.signingSecret.toString(),
      },
    });

    if (this.eventBus) {
      handler.addEnvironment('EVENT_BUS_NAME', this.eventBus.eventBusName);
    }

    events.EventBus.grantAllPutEvents(handler);

    // HTTP API
    const httpApi = new apigatewayv2.HttpApi(this, 'SlackEventsApi', {
      defaultIntegration: new integrations.HttpLambdaIntegration('Integration', handler),
      apiName: props.apiName,
    });

    new cdk.CfnOutput(this, 'ApiEndpoint', {
      value: httpApi.apiEndpoint,
    });
  }
}
