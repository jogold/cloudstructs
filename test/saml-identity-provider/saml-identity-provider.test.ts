import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as iam from 'aws-cdk-lib/aws-iam';
import { SamlFederatedPrincipal, SamlIdentityProvider } from '../../src';

let stack: Stack;
beforeEach(() => {
  stack = new Stack();
});

test('SamlIdentityProvider', () => {
  const identityProvider = new SamlIdentityProvider(stack, 'IdentityProvider', {
    metadataDocument: '<?xml version="1.0" encoding="utf-8"?><EntityDescriptor ID="ID" entityID="ID"</EntityDescriptor>',
  });

  new iam.Role(stack, 'Role', {
    assumedBy: new SamlFederatedPrincipal(identityProvider),
  });

  expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
});
