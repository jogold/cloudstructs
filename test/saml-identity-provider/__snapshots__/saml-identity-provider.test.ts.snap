// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SamlIdentityProvider 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "11d371859e49f3b1a49896583254890801a40beada56e4c84409a179a221e2e2.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "IdentityProviderA00909EA": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "IdentityProviderCustomResourcePolicy3F0F7442",
      ],
      "Properties": {
        "Create": "{"service":"IAM","action":"createSAMLProvider","parameters":{"Name":"IdentityProviderIdentityProvider","SAMLMetadataDocument":"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?><EntityDescriptor ID=\\"ID\\" entityID=\\"ID\\"</EntityDescriptor>"},"physicalResourceId":{"responsePath":"SAMLProviderArn"}}",
        "Delete": "{"service":"IAM","action":"deleteSAMLProvider","parameters":{"SAMLProviderArn":"PHYSICAL:RESOURCEID:"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": "{"service":"IAM","action":"updateSAMLProvider","parameters":{"SAMLProviderArn":"PHYSICAL:RESOURCEID:","SAMLMetadataDocument":"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?><EntityDescriptor ID=\\"ID\\" entityID=\\"ID\\"</EntityDescriptor>"},"physicalResourceId":{"responsePath":"SAMLProviderArn"}}",
      },
      "Type": "Custom::SamlIdentityProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "IdentityProviderCustomResourcePolicy3F0F7442": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "iam:createSAMLProvider",
                "iam:updateSAMLProvider",
                "iam:deleteSAMLProvider",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iam::",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":saml-provider/IdentityProviderIdentityProvider",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "IdentityProviderCustomResourcePolicy3F0F7442",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Role1ABCC5F0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithSAML",
              "Condition": {
                "StringEquals": {
                  "SAML:aud": "https://signin.aws.amazon.com/saml",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Fn::GetAtt": [
                    "IdentityProviderA00909EA",
                    "SAMLProviderArn",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
